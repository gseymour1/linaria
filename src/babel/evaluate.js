/* @flow */

import type { Options as PluginOptions } from './extract';

const generator = require('@babel/generator').default;
const babel = require('@babel/core');
const Module = require('./module');

const resolve = (path, t, requirements) => {
  const binding = path.scope.getBinding(path.node.name);

  if (
    path.isReferenced() &&
    binding &&
    binding.kind !== 'param' &&
    !requirements.some(req => req.path === binding.path)
  ) {
    let result;

    switch (binding.kind) {
      case 'module':
        if (t.isImportSpecifier(binding.path)) {
          result = t.importDeclaration(
            [binding.path.node],
            binding.path.parentPath.node.source
          );
        } else {
          result = binding.path.parentPath.node;
        }
        break;
      case 'const':
      case 'let':
      case 'var': {
        result = t.variableDeclaration(binding.kind, [binding.path.node]);
        break;
      }
      default:
        result = binding.path.node;
        break;
    }

    const { loc } = binding.path.node;

    requirements.push({
      result,
      path: binding.path,
      start: loc.start,
      end: loc.end,
    });

    binding.path.traverse({
      Identifier(p) {
        resolve(p, t, requirements);
      },
    });
  }
};

module.exports = function evaluate(
  path: any,
  t: any,
  filename: string,
  transformer?: (text: string) => { code: string },
  options?: PluginOptions
) {
  const requirements = [];

  if (t.isIdentifier(path)) {
    resolve(path, t, requirements);
  } else {
    path.traverse({
      Identifier(p) {
        resolve(p, t, requirements);
      },
    });
  }

  // Collect the list of dependencies that we import
  const dependencies = requirements.reduce((deps, req) => {
    if (t.isImportDeclaration(req.path.parentPath)) {
      deps.push(req.path.parentPath.node.source.value);
    } else {
      req.path.traverse({
        CallExpression(p) {
          const { callee, arguments: args } = p.node;

          let name;

          if (callee.name === 'require' && args.length === 1) {
            if (
              args[0].type === 'Literal' ||
              args[0].type === 'StringLiteral'
            ) {
              name = args[0].value;
            } else if (
              args[0].type === 'TemplateLiteral' &&
              args[0].quasis.length === 1
            ) {
              name = args[0].quasis[0].value.cooked;
            }
          }

          if (name) {
            deps.push(name);
          }
        },
      });
    }

    return deps;
  }, []);

  const expression = t.expressionStatement(
    t.assignmentExpression(
      '=',
      t.memberExpression(t.identifier('module'), t.identifier('exports')),
      path.node
    )
  );

  // Preserve source order
  requirements.sort((a, b) => {
    if (a.start.line === b.start.line) {
      return a.start.column - b.start.column;
    }

    return a.start.line - b.start.line;
  });

  // We'll wrap each code in a block to avoid collisions in variable names
  // We separate out the imports since they cannot be inside blocks
  const { imports, others } = requirements.reduce(
    (acc, curr) => {
      if (t.isImportDeclaration(curr.path.parentPath)) {
        acc.imports.push(curr.result);
      } else {
        // Add these in reverse because we'll need to wrap in block statements in reverse
        acc.others.unshift(curr.result);
      }

      return acc;
    },
    { imports: [], others: [] }
  );

  const wrapped = others.reduce(
    (acc, curr) => t.blockStatement([curr, acc]),
    t.blockStatement([expression])
  );

  const m = new Module(filename);

  m.transform =
    typeof transformer !== 'undefined'
      ? transformer
      : function transform(text) {
          if (options && options.ignore && options.ignore.test(this.filename)) {
            return { code: text };
          }

          return babel.transformSync(text, {
            caller: { name: 'linaria', evaluate: true },
            filename: this.filename,
            presets: [[require.resolve('./index'), options]],
            plugins: [
              // Include this plugin to avoid extra config when using { module: false } for webpack
              '@babel/plugin-transform-modules-commonjs',
              '@babel/plugin-proposal-export-namespace-from',
              // We don't support dynamic imports when evaluating, but don't wanna syntax error
              // This will replace dynamic imports with an object that does nothing
              require.resolve('./dynamic-import-noop'),
            ],
          });
        };

  m.evaluate(
    [
      // Use String.raw to preserve escapes such as '\n' in the code
      // Flow doesn't understand template tags: https://github.com/facebook/flow/issues/2616
      /* $FlowFixMe */
      imports.map(node => String.raw`${generator(node).code}`).join('\n'),
      /* $FlowFixMe */
      String.raw`${generator(wrapped).code}`,
    ].join('\n')
  );

  return {
    value: m.exports,
    dependencies,
  };
};
