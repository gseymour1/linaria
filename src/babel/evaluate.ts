// TypeScript Version: 3.2

import {
  transformSync,
  BabelFileResult,
  PluginItem,
  TransformOptions,
  types,
} from '@babel/core';
import generator from '@babel/generator';
import Module from './module';
import { Location, StrictOptions, Value } from './types';
import { NodePath } from '@babel/traverse';

type DefaultOptions = Partial<TransformOptions> & {
  plugins: PluginItem[];
  presets: PluginItem[];
};

interface IRequirement {
  result: types.Node;
  path: NodePath;
  start: Location;
  end: Location;
}

const isAdded = (requirements: IRequirement[], path: NodePath): boolean => {
  if (requirements.some(req => req.path === path)) {
    return true;
  }

  if (path.parentPath) {
    return isAdded(requirements, path.parentPath);
  }

  return false;
};

const resolve = (
  path: NodePath<types.Identifier>,
  requirements: IRequirement[]
) => {
  const binding = path.scope.getBinding(path.node.name);

  if (
    path.isReferenced() &&
    binding &&
    // Next condition it's always true because `params` isn't valid value
    (binding.kind as string) !== 'param' &&
    !isAdded(requirements, binding.path)
  ) {
    let result;

    switch (binding.kind) {
      case 'module':
        if (types.isImportSpecifier(binding.path)) {
          const p = binding.path as NodePath<types.ImportSpecifier>;
          result = types.importDeclaration(
            [p.node],
            (p.parentPath.node as types.ImportDeclaration).source
          );
        } else {
          result = binding.path.parentPath.node;
        }
        break;
      case 'const':
      case 'let':
      case 'var': {
        const { node } = binding.path as NodePath<types.VariableDeclarator>;
        let decl;

        // Replace SequenceExpressions (expr1, expr2, expr3, ...) with the last one
        if (types.isSequenceExpression(node.init)) {
          decl = types.variableDeclarator(
            node.id,
            node.init.expressions[node.init.expressions.length - 1]
          );
        } else {
          decl = node;
        }

        result = types.variableDeclaration(binding.kind, [decl]);
        break;
      }
      default:
        result = binding.path.node;
        break;
    }

    const { loc } = binding.path.node;

    requirements.push({
      result,
      path: binding.path,
      start: loc!.start,
      end: loc!.end,
    });

    binding.path.traverse({
      Identifier(p) {
        resolve(p, requirements);
      },
    });
  }
};

export default function evaluate(
  path: any,
  t: any,
  filename: string,
  transformer?: (text: string) => BabelFileResult | null,
  options?: StrictOptions
) {
  if (t.isSequenceExpression(path)) {
    // We only need to evaluate the last item in a sequence expression, e.g. (a, b, c)
    // eslint-disable-next-line no-param-reassign
    path = path.get('expressions')[path.node.expressions.length - 1];
  }

  const requirements: IRequirement[] = [];

  if (t.isIdentifier(path)) {
    resolve(path, requirements);
  } else {
    path.traverse({
      Identifier(p: NodePath<types.Identifier>) {
        resolve(p, requirements);
      },
    });
  }

  const expression = t.expressionStatement(
    t.assignmentExpression(
      '=',
      t.memberExpression(t.identifier('module'), t.identifier('exports')),
      path.node
    )
  );

  // Preserve source order
  requirements.sort((a, b) => {
    if (a.start.line === b.start.line) {
      return a.start.column - b.start.column;
    }

    return a.start.line - b.start.line;
  });

  // We'll wrap each code in a block to avoid collisions in variable names
  // We separate out the imports since they cannot be inside blocks
  const { imports, others } = requirements.reduce(
    (acc, curr) => {
      if (t.isImportDeclaration(curr.path.parentPath)) {
        acc.imports.push(curr.result);
      } else {
        // Add these in reverse because we'll need to wrap in block statements in reverse
        acc.others.unshift(curr.result);
      }

      return acc;
    },
    { imports: [] as types.Node[], others: [] as types.Node[] }
  );

  const wrapped = others.reduce(
    (acc, curr) => t.blockStatement([curr, acc]),
    t.blockStatement([expression])
  );

  const m = new Module(filename);

  m.dependencies = [];
  m.transform =
    typeof transformer !== 'undefined'
      ? transformer
      : function transform(this: Module, text) {
          if (options && options.ignore && options.ignore.test(this.filename)) {
            return { code: text };
          }

          const plugins: Array<string | object> = [
            // Include these plugins to avoid extra config when using { module: false } for webpack
            '@babel/plugin-transform-modules-commonjs',
            '@babel/plugin-proposal-export-namespace-from',
          ];

          const defaults: DefaultOptions = {
            caller: { name: 'linaria', evaluate: true },
            filename: this.filename,
            presets: [[require.resolve('./index'), options]],
            plugins: [
              ...plugins.map(name => require.resolve(name as string)),
              // We don't support dynamic imports when evaluating, but don't wanna syntax error
              // This will replace dynamic imports with an object that does nothing
              require.resolve('./dynamic-import-noop'),
            ],
          };

          const babelOptions =
            // Shallow copy the babel options because we mutate it later
            options && options.babelOptions ? { ...options.babelOptions } : {};

          // If we programmtically pass babel options while there is a .babelrc, babel might throw
          // We need to filter out duplicate presets and plugins so that this doesn't happen
          // This workaround isn't full proof, but it's still better than nothing
          const keys: Array<
            keyof TransformOptions & ('presets' | 'plugins')
          > = ['presets', 'plugins'];
          keys.forEach(field => {
            babelOptions[field] = babelOptions[field]
              ? babelOptions[field]!.filter((item: PluginItem) => {
                  // If item is an array it's a preset/plugin with options ([preset, options])
                  // Get the first item to get the preset.plugin name
                  // Otheriwse it's a plugin name (can be a function too)
                  const name = Array.isArray(item) ? item[0] : item;

                  if (
                    // In our case, a preset might also be referring to linaria/babel
                    // We require the file from internal path which is not the same one that we export
                    // This case won't get caught and the preset won't filtered, even if they are same
                    // So we add an extra check for top level linaria/babel
                    name === 'linaria/babel' ||
                    name === require.resolve('../../babel') ||
                    // Also add a check for the plugin names we include for bundler support
                    plugins.includes(name)
                  ) {
                    return false;
                  }

                  // Loop through the default presets/plugins to see if it already exists
                  return !defaults[field].some(it =>
                    // The default presets/plugins can also have nested arrays,
                    Array.isArray(it) ? it[0] === name : it === name
                  );
                })
              : [];
          });

          return transformSync(text, {
            // Passed options shouldn't be able to override the options we pass
            // Linaria's plugins rely on these (such as filename to generate consistent hash)
            ...babelOptions,
            ...defaults,
            presets: [
              // Preset order is last to first, so add the extra presets to start
              // This makes sure that our preset is always run first
              ...babelOptions.presets!,
              ...defaults.presets,
            ],
            plugins: [
              ...defaults.plugins,
              // Plugin order is first to last, so add the extra presets to end
              // This makes sure that the plugins we specify always run first
              ...babelOptions.plugins!,
            ],
          });
        };

  m.evaluate(
    [
      // Use String.raw to preserve escapes such as '\n' in the code
      // Flow doesn't understand template tags: https://github.com/facebook/flow/issues/2616
      /* $FlowFixMe */
      imports.map(node => String.raw`${generator(node).code}`).join('\n'),
      /* $FlowFixMe */
      String.raw`${generator(wrapped).code}`,
    ].join('\n')
  );

  return {
    value: m.exports as Value,
    dependencies: m.dependencies,
  };
}
