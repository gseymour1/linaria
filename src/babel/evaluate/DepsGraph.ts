import { types as t } from '@babel/core';
import ScopeManager from './scope';

export type ExternalDep = {
  source: string;
  local: t.Identifier;
  imported: t.Identifier | null;
};

export default class DepsGraph {
  public readonly externalDeps: ExternalDep[] = [];

  private readonly edges: Array<[t.Node, t.Node]> = [];
  private readonly dependencies: Map<t.Node, t.Node[]> = new Map();
  private readonly dependents: Map<t.Node, t.Node[]> = new Map();
  private readonly bindingsDependencies: Map<string, t.Node[]> = new Map();
  private readonly bindingsDependents: Map<string, t.Node[]> = new Map();

  constructor(private scope: ScopeManager) {}

  addEdge(a: t.Node, b: t.Node) {
    this.edges.push([a, b]);
    if (!this.dependencies.has(a)) {
      this.dependencies.set(a, []);
    }

    if (!this.dependents.has(b)) {
      this.dependents.set(b, []);
    }

    this.dependencies.get(a)!.push(b);
    this.dependents.get(b)!.push(a);

    if (t.isIdentifier(a)) {
      const scopeId = this.scope.whereIsDeclared(a);
      if (scopeId !== undefined) {
        const key = `${scopeId}:${a.name}`;
        if (!this.bindingsDependencies.has(key)) {
          this.bindingsDependencies.set(key, []);
        }

        this.bindingsDependencies.get(key)!.push(b);
      }
    }

    if (t.isIdentifier(b)) {
      const scopeId = this.scope.whereIsDeclared(b);
      if (scopeId !== undefined) {
        const key = `${scopeId}:${b.name}`;
        if (!this.bindingsDependents.has(key)) {
          this.bindingsDependents.set(key, []);
        }

        this.bindingsDependents.get(key)!.push(a);
      }
    }
  }

  getDependenciesByBinding(id: string) {
    return this.bindingsDependencies.get(id) || [];
  }

  getDependentsByBinding(id: string) {
    return this.bindingsDependents.get(id) || [];
  }

  findDependencies(like: Object) {
    return this.edges
      .filter(([a]) => t.shallowEqual(a, like))
      .map(([, b]) => b);
  }

  findDependents(like: object) {
    return this.edges
      .filter(([, b]) => t.shallowEqual(b, like))
      .map(([a]) => a);
  }

  getDependencies(nodes: t.Node[]) {
    return nodes.reduce(
      (acc, node) => acc.concat(this.dependencies.get(node) || []),
      [] as t.Node[]
    );
  }

  getDependents(nodes: t.Node[]) {
    return nodes.reduce(
      (acc, node) => acc.concat(this.dependents.get(node) || []),
      [] as t.Node[]
    );
  }

  getLeafs(required: Array<t.Expression | string>): t.Expression[] {
    return required
      .map(node =>
        typeof node === 'string' ? this.scope.getDeclaration(`0:${node}`) : node
      )
      .filter(n => n) as t.Expression[];
  }

  isDeclared(name: string) {
    return this.scope.isDeclared(`0:${name}`);
  }
}
