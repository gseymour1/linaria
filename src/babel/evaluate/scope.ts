import { types } from '@babel/core';

type Identifier = types.Identifier;
type Scope = Map<string, Set<Identifier>>;

export type ScopeId = number | 'global';
export type DeclareHandler = (identifier: Identifier) => void;

const scopeIds = new WeakMap<Scope, number | 'global'>();
const getId = (scope: Scope, identifier: Identifier): string =>
  `${scopeIds.get(scope)}:${identifier.name}`;

export default class ScopeManager {
  private nextId = 0;
  private readonly stack: Array<Scope> = [];
  private readonly map: Map<ScopeId, Scope> = new Map();
  private readonly handlers: Map<ScopeId, Array<DeclareHandler>> = new Map();
  private readonly declarations: Map<string, Identifier> = new Map();

  private get global(): Scope {
    return this.map.get('global')!;
  }

  constructor() {
    this.new('global');
  }

  new(scopeId: ScopeId = this.nextId++): Scope {
    const scope: Scope = new Map();
    scopeIds.set(scope, scopeId);
    this.map.set(scopeId, scope);
    this.handlers.set(scopeId, []);
    this.stack.unshift(scope);
    return scope;
  }

  dispose(): Scope | undefined {
    const disposed = this.stack.shift();
    if (disposed) {
      this.map.delete(scopeIds.get(disposed)!);
    }

    return disposed;
  }

  declare(identifier: Identifier): void {
    const scope = this.stack[0];
    scope.set(identifier.name, new Set([identifier]));
    this.declarations.set(getId(scope, identifier), identifier);
    const handlers = this.handlers.get(scopeIds.get(scope)!)!;
    handlers.forEach(handler => handler(identifier));
  }

  addReference(identifier: Identifier): Identifier | undefined {
    const name = identifier.name;
    const scope = this.stack.find(s => s.has(name)) || this.global;
    const id = getId(scope, identifier);
    if (scope === this.global && !scope.has(name)) {
      scope.set(name, new Set());
    }

    scope.get(name)!.add(identifier);
    return this.declarations.get(id);
  }

  whereIsDeclared(identifier: Identifier): ScopeId | undefined {
    const name = identifier.name;
    const scope = this.stack.find(
      s => s.has(name) && s.get(name)!.has(identifier)
    );
    if (scope) {
      return scopeIds.get(scope);
    }

    if (this.global.has(name)) {
      return 'global';
    }

    return undefined;
  }

  isDeclared(name: string): boolean {
    return this.declarations.has(name);
  }

  getDeclaration(
    identifierOrName: Identifier | string
  ): Identifier | undefined {
    let name: string;
    if (typeof identifierOrName === 'string') {
      name = identifierOrName;
    } else {
      const scopeId = this.whereIsDeclared(identifierOrName);
      if (scopeId === undefined || scopeId === 'global') {
        return undefined;
      }

      name = getId(this.map.get(scopeId)!, identifierOrName);
    }

    return this.declarations.get(name);
  }

  addDeclareHandler(handler: DeclareHandler): () => void {
    const scopeId = scopeIds.get(this.stack[0])!;
    this.handlers.get(scopeId)!.push(handler);
    return () => {
      const handlers = this.handlers.get(scopeId)!.filter(h => h !== handler);
      this.handlers.set(scopeId, handlers);
    };
  }
}
