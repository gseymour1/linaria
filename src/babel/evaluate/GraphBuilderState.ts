import { types as t } from '@babel/core';
import ScopeManager from './scope';
import DepsGraph from './DepsGraph';

export default abstract class GraphBuilderState {
  public readonly scope = new ScopeManager();
  public readonly graph = new DepsGraph(this.scope);
  public readonly meta = new Map<string, any>();

  /*
   * For expressions like `{ foo: bar }` we need to now context
   *
   * const obj = { foo: bar };
   * Here context is `expression`, `bar` is a variable which depends from its declaration.
   *
   * const { foo: bar } = obj;
   * Here context is `pattern` and `bar` is a variable declaration itself.
   */
  public readonly context: Array<'expression' | 'pattern' | 'lval'> = [];

  public readonly fnStack: t.Node[] = [];

  abstract baseVisit<TNode extends t.Node>(
    node: TNode,
    ignoreDeps?: boolean
  ): void;

  abstract visit<TNode extends t.Node, TParent extends t.Node>(
    node: TNode,
    parent: TParent | null,
    parentKey: t.VisitorKeys[TParent['type']] | null,
    listIdx?: number | null
  ): void;
}
