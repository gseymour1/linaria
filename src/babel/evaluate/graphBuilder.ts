import { types as t } from '@babel/core';
import isNode from '../utils/isNode';
import getVisitorKeys from '../utils/getVisitorKeys';
import DepsGraph from './DepsGraph';
import GraphBuilderState from './GraphBuilderState';
import { getVisitor } from './Visitors';

class GraphBuilder extends GraphBuilderState {
  static build(root: t.Node): DepsGraph {
    const builder = new GraphBuilder();
    builder.visit(root, null, null, null);
    return builder.graph;
  }

  baseVisit<TNode extends t.Node>(node: TNode, ignoreDeps = false) {
    const dependencies = [];
    const isExpression = t.isExpression(node);
    const keys = getVisitorKeys(node);
    for (const key of keys) {
      const subNode = node[key as keyof TNode];

      if (Array.isArray(subNode)) {
        for (let i = 0; i < subNode.length; i++) {
          const child = subNode[i];
          if (child) {
            this.visit(child, node, key, i);
            dependencies.push(child);
          }
        }
      } else if (isNode(subNode)) {
        this.visit(subNode, node, key);
        dependencies.push(subNode);
      }
    }

    if (isExpression && !ignoreDeps) {
      dependencies.forEach(dep => this.graph.addEdge(node, dep));
    }
  }

  visit<TNode extends t.Node, TParent extends t.Node>(
    node: TNode,
    parent: TParent | null,
    parentKey: t.VisitorKeys[TParent['type']] | null,
    listIdx: number | null = null
  ) {
    const isScopable = t.isScopable(node);
    const isFunction = t.isFunction(node);

    if (isScopable) this.scope.new();
    if (isFunction) this.fnStack.push(node);

    const visitor = getVisitor(node);
    if (visitor) {
      visitor.call(this, node, parent, parentKey, listIdx);
    } else {
      this.baseVisit(node);
    }

    if (isFunction) this.fnStack.pop();
    if (isScopable) this.scope.dispose();
  }
}

export default GraphBuilder.build;
