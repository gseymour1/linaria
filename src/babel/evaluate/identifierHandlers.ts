import { types as t } from '@babel/core';
import GraphBuilderState from './GraphBuilderState';
import peek from '../utils/peek';

type HandlerFn = <TParent extends t.Node = t.Node>(
  builder: GraphBuilderState,
  node: t.Identifier,
  parent: TParent,
  parentKey: t.VisitorKeys[TParent['type']],
  listIdx: number | null
) => void;

type Handler = 'declare' | 'keep' | 'refer' | HandlerFn;

type NodeType = t.Node['type'] | keyof t.Aliases;

const handlers: {
  [key: string]: Handler;
} = {};

function isAlias(type: NodeType): type is keyof t.Aliases {
  return type in t.FLIPPED_ALIAS_KEYS;
}

export function defineHandler<T extends NodeType>(
  typeOrAlias: T,
  field: string,
  handler: Handler
) {
  const types = isAlias(typeOrAlias)
    ? t.FLIPPED_ALIAS_KEYS[typeOrAlias]
    : [typeOrAlias];
  types.forEach((type: string) => {
    handlers[`${type}:${field}`] = handler;
  });
}

export function batchDefineHandlers(
  typesAndFields: [NodeType, string][],
  handler: Handler
) {
  typesAndFields.forEach(([type, field]) =>
    defineHandler(type, field, handler)
  );
}

batchDefineHandlers(
  [
    ['ArrayPattern', 'elements'],
    ['AssignmentPattern', 'left'],
    ['Function', 'params'],
    ['ImportDefaultSpecifier', 'local'],
    ['ImportSpecifier', 'local'],
    ['RestElement', 'argument'],
    ['VariableDeclarator', 'id'],
  ],
  'declare'
);

batchDefineHandlers(
  [
    ['ExportSpecifier', 'exported'],
    ['ImportSpecifier', 'imported'],
    ['MemberExpression', 'property'],
    ['ObjectProperty', 'key'],
  ],
  'keep'
);

batchDefineHandlers(
  [
    ['AssignmentExpression', 'left'],
    ['AssignmentExpression', 'right'],
    ['AssignmentPattern', 'right'],
    ['BinaryExpression', 'left'],
    ['BinaryExpression', 'right'],
    ['CallExpression', 'arguments'],
    ['CallExpression', 'callee'],
    ['ExportDefaultDeclaration', 'declaration'],
    ['ExportSpecifier', 'local'],
    ['Function', 'body'],
    ['LogicalExpression', 'left'],
    ['LogicalExpression', 'right'],
    ['MemberExpression', 'property'],
    ['MemberExpression', 'object'],
    ['NewExpression', 'callee'],
    ['ReturnStatement', 'argument'],
    ['SequenceExpression', 'expressions'],
    ['SpreadElement', 'argument'],
    ['TaggedTemplateExpression', 'tag'],
    ['TemplateLiteral', 'expressions'],
    ['VariableDeclarator', 'init'],
  ],
  'refer'
);

/*
 * Special handler for { foo: bar } in different contexts
 */
defineHandler(
  'ObjectProperty',
  'value',
  (builder: GraphBuilderState, node: t.Identifier) => {
    const context = peek(builder.context);
    if (context === 'pattern') {
      builder.scope.declare(node);
    } else {
      const declaration = builder.scope.addReference(node);
      if (declaration) {
        builder.graph.addEdge(node, declaration);
      }
    }
  }
);

export default handlers;
