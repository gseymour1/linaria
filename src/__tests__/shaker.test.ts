/* eslint-disable no-template-curly-in-string */

import dedent from 'dedent';
import * as babel from '@babel/core';
import { ExternalDep } from '../babel/evaluate/DepsGraph';
import shake from '../babel/evaluate/shaker';

function _build(literal: TemplateStringsArray, ...placeholders: string[]) {
  const code = dedent(literal, ...placeholders);
  return {
    ast: babel.parseSync(code),
    code,
  };
}

function _shake(names: string[]) {
  return (
    literal: TemplateStringsArray,
    ...placeholders: string[]
  ): [string, ExternalDep[]] => {
    const { ast, code } = _build(literal, ...placeholders);
    if (!ast || !babel.types.isFile(ast)) return ['', []];
    const [shaken, deps] = shake(ast.program, names);
    const { code: transformed } = babel.transformFromAstSync(shaken, code)!;
    return [transformed!, deps];
  };
}

it('removes all', () => {
  const [shaken] = _shake([])`
    const { whiteColor: color, anotherColor } = require('…');
    const a = color || anotherColor;
    color.green = '#0f0';
  `;

  expect(shaken).toBe('module.exports = [];');
});

it('keeps only code which is related to `color`', () => {
  const [shaken] = _shake(['color'])`
    const { whiteColor: color, anotherColor } = require('…');
    const wrap = '';
    const a = color || anotherColor;
    color.green = '#0f0';
    module.exports = { color, anotherColor };
  `;

  expect(shaken).toMatchSnapshot();
});

it('keeps only code which is related to `anotherColor`', () => {
  const [shaken] = _shake(['anotherColor'])`
    const { whiteColor: color, anotherColor } = require('…');
    const a = color || anotherColor;
    color.green = '#0f0';
  `;

  expect(shaken).toMatchSnapshot();
});

it('keeps only code which is related to `a`', () => {
  const [shaken] = _shake(['a'])`
    const { whiteColor: color, anotherColor } = require('…');
    const a = color || anotherColor;
    color.green = '#0f0';
  `;

  expect(shaken).toMatchSnapshot();
});

it('shakes imports', () => {
  const [shaken] = _shake(['color', 'defaultColor'])`
    import { unrelatedImport } from '…';
    import { whiteColor as color, anotherColor } from '…';
    import defaultColor from '…';
    import anotherDefaultColor from '…';
    import '…';
    require('…');
    export default color;
  `;

  expect(shaken).toMatchSnapshot();
});

it('should keep member expression key', () => {
  const [shaken] = _shake(['blue'])`
    const key = 'blue';
    const obj = { blue: '#00F' };
    const blue = obj[key];
  `;

  expect(shaken).toMatchSnapshot();
});

it('shakes exports', () => {
  const [shaken] = _shake(['a'])`
    import { whiteColor as color, anotherColor } from '…';
    export const a = color;
    export { anotherColor };
  `;

  expect(shaken).toMatchSnapshot();
});

it('should throw away any side effects', () => {
  const [shaken] = _shake(['foo'])`
    const objects = { key: { fontSize: 12 } };
    const foo = (k) => {
      const obj = objects[k];
      console.log('side effect');
      return obj;
    };
  `;

  expect(shaken).toMatchSnapshot();
});

it('keeps objects as is', () => {
  const [shaken] = _shake(['fill1', 'fill2'])`
    const fill1 = (top = 0, left = 0, right = 0, bottom = 0) => ({
      position: 'absolute',
      top,
      right,
      bottom,
      left,
    });

    const fill2 = (top = 0, left = 0, right = 0, bottom = 0) => {
      return {
        position: 'absolute',
        top,
        right,
        bottom,
        left,
      };
    };
  `;

  expect(shaken).toMatchSnapshot();
});

it('shakes sequence expression', () => {
  const [shaken] = _shake(['color2'])`
    import { external } from '…';
    const color1 = (external, () => 'blue');
    let local = '';
    const color2 = (local = color1(), () => local);
  `;

  expect(shaken).toMatchSnapshot();
});
