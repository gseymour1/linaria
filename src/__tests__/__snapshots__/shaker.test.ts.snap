// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keeps objects as is 1`] = `
"const fill1 = (top = 0, left = 0, right = 0, bottom = 0) => ({
  position: 'absolute',
  top,
  right,
  bottom,
  left
});

const fill2 = (top = 0, left = 0, right = 0, bottom = 0) => {
  return {
    position: 'absolute',
    top,
    right,
    bottom,
    left
  };
};

const wrap = fn => {
  try {
    return fn();
  } catch (e) {
    return e;
  }
};

module.exports = [wrap(() => fill1), wrap(() => fill2)];"
`;

exports[`keeps only code which is related to \`a\` 1`] = `
"const {
  whiteColor: color,
  anotherColor
} = require('…');

const a = color || anotherColor;
color.green = '#0f0';

const wrap = fn => {
  try {
    return fn();
  } catch (e) {
    return e;
  }
};

module.exports = [wrap(() => a)];"
`;

exports[`keeps only code which is related to \`anotherColor\` 1`] = `
"const {
  anotherColor
} = require('…');

const wrap = fn => {
  try {
    return fn();
  } catch (e) {
    return e;
  }
};

module.exports = [wrap(() => anotherColor)];"
`;

exports[`keeps only code which is related to \`color\` 1`] = `
"const {
  whiteColor: color
} = require('…');

color.green = '#0f0';

const wrap_1 = fn => {
  try {
    return fn();
  } catch (e) {
    return e;
  }
};

module.exports = [wrap_1(() => color)];"
`;

exports[`shakes exports 1`] = `
"import { whiteColor as color } from '…';
const a = color;

const wrap = fn => {
  try {
    return fn();
  } catch (e) {
    return e;
  }
};

module.exports = [wrap(() => a)];"
`;

exports[`shakes imports 1`] = `
"import { whiteColor as color } from '…';
import defaultColor from '…';

const wrap = fn => {
  try {
    return fn();
  } catch (e) {
    return e;
  }
};

module.exports = [wrap(() => color), wrap(() => defaultColor)];"
`;

exports[`shakes sequence expression 1`] = `
"const color1 = (() => 'blue');
let local = '';
const color2 = (local = color1(), () => local);

const wrap = fn => {
  try {
    return fn();
  } catch (e) {
    return e;
  }
};

module.exports = [wrap(() => color2)];"
`;

exports[`should keep member expression key 1`] = `
"const key = 'blue';
const obj = {
  blue: '#00F'
};
const blue = obj[key];

const wrap = fn => {
  try {
    return fn();
  } catch (e) {
    return e;
  }
};

module.exports = [wrap(() => blue)];"
`;

exports[`should throw away any side effects 1`] = `
"const objects = {
  key: {
    fontSize: 12
  }
};

const foo = k => {
  const obj = objects[k];
  return obj;
};

const wrap = fn => {
  try {
    return fn();
  } catch (e) {
    return e;
  }
};

module.exports = [wrap(() => foo)];"
`;
